File: 724-Find-Pivot-Index/main.go
package main

// my solution
// T: O(n)
// M: O(n)
func pivotIndex(nums []int) int {
	leftPrefixSum := make([]int, 0, len(nums))
	rightPrefixSum := make([]int, 0, len(nums))

	for i := 0; i < len(nums); i++ {
		leftPrefixSum = append(leftPrefixSum, 0)
		rightPrefixSum = append(rightPrefixSum, 0)
	}

	prefix := 0

	for i := 1; i < len(nums); i++ {
		prefix += nums[i-1]
		leftPrefixSum[i] = prefix
	}

	prefix = 0

	for i := len(nums) - 2; i >= 0; i-- {
		prefix += nums[i+1]
		rightPrefixSum[i] = prefix
	}

	for i := 0; i < len(nums); i++ {
		if leftPrefixSum[i] == rightPrefixSum[i] {
			return i
		}
	}

	return -1
}

File: 724-Find-Pivot-Index/main.py
from typing import List


# 1. Brute force
# Go through each el. For each el, calculate it's left and right sum and compare them.

# T: O(n^2)
# M: O(1)
class Solution:
    def pivotIndex(self, nums: List[int]) -> int:


# T: O(n)
# M: O(1)
class Solution:
    def pivotIndex(self, nums: List[int]) -> int:
        total = sum(nums) # T: O(n)
        leftSum = 0

        # T: O(n)
        for i in range(len(nums)):
            rightSum = total - leftSum - nums[i]

            if leftSum == rightSum:
                return i

            # in the next iteration, curr el is added to the leftSum:
            leftSum += nums[i]

        return -1